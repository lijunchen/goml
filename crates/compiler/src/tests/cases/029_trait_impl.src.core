fn impl_Display_Point_show(self/0: Point) -> string {
  "Point"
}

fn impl_Display_Option_int_show(self/1: Option[int]) -> string {
  match self/1 {
    Option::Some(x0) => {
      let x0 = Option::Some._0(self/1) in
      let value/2 = x0 in
      string_add("Some(", string_add(int_to_string(value/2), ")"))
    },
    Option::None => {
      "None"
    },
  }
}

fn make_optional(flag/3: bool) -> Option[int] {
  if flag/3 {
      Option::Some(42)
  } else {
      Option::None
  }
}

fn main() -> unit {
  let point/4 = Point { x: 1, y: 2 } in
  let some_number/5 = make_optional(true) in
  let none_number/6 = make_optional(false) in
  let mtmp1 = string_println(impl_Display_Point_show(point/4)) in
  let mtmp2 = string_println(impl_Display_Option_int_show(some_number/5)) in
  let mtmp3 = string_println(impl_Display_Option_int_show(none_number/6)) in
  ()
}