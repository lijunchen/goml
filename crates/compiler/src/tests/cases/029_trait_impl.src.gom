package main

import (
    "fmt"
)

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func string_add(a string, b string) string {
    return a + b
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

type Point struct {
    x int
    y int
}

type Option__int interface {
    isOption__int()
}

type Some struct {
    _0 int
}

func (_ Some) isOption__int() {}

type None struct {}

func (_ None) isOption__int() {}

func impl_Display_Point_show(self__0 Point) string {
    var ret9 string
    ret9 = "Point"
    return ret9
}

func impl_Display_Option_int_show(self__1 Option__int) string {
    var ret10 string
    switch self__1 := self__1.(type) {
    case Some:
        var x0 int = self__1._0
        var value__2 int = x0
        var t5 string = int_to_string(value__2)
        var t4 string = string_add(t5, ")")
        ret10 = string_add("Some(", t4)
    case None:
        ret10 = "None"
    }
    return ret10
}

func make_optional(flag__3 bool) Option__int {
    var ret11 Option__int
    if flag__3 {
        ret11 = Some{
            _0: 42,
        }
    } else {
        ret11 = None{}
    }
    return ret11
}

func main0() struct{} {
    var ret12 struct{}
    var point__4 Point = Point{
        x: 1,
        y: 2,
    }
    var some_number__5 Option__int = make_optional(true)
    var none_number__6 Option__int = make_optional(false)
    var t6 string = impl_Display_Point_show(point__4)
    string_println(t6)
    var t7 string = impl_Display_Option_int_show(some_number__5)
    string_println(t7)
    var t8 string = impl_Display_Option_int_show(none_number__6)
    string_println(t8)
    ret12 = struct{}{}
    return ret12
}

func main() {
    main0()
}
