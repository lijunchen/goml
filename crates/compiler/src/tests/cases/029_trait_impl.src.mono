fn impl_Display_Point_show(self/0: Point) -> string {
  "Point"
}

fn impl_Display_Option_int_show(self/1: Option__int) -> string {
  match self/1 {
    Option__int::Some(x0) => {
      let x0 = Option__int::Some._0(self/1) in
      let value/2 = x0 in
      string_add("Some(", string_add(int_to_string(value/2), ")"))
    },
    Option__int::None => {
      "None"
    },
  }
}

fn make_optional(flag/3: bool) -> Option__int {
  if flag/3 {
      Option__int::Some(42)
  } else {
      Option__int::None
  }
}

fn main() -> unit {
  let point/4 = Point { x: 1, y: 2 } in
  let some_number/5 = make_optional(true) in
  let none_number/6 = make_optional(false) in
  let mtmp1 = string_println(impl_Display_Point_show(point/4)) in
  let mtmp2 = string_println(impl_Display_Option_int_show(some_number/5)) in
  let mtmp3 = string_println(impl_Display_Option_int_show(none_number/6)) in
  ()
}