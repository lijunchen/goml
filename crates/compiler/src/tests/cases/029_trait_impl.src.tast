impl Display for Point{
  fn show(self/0: Point) -> string {
    "Point"
  }
}

impl Display for Option[int]{
  fn show(self/1: Option[int]) -> string {
    match (self/1 : Option[int]) {
        Option::Some(value/2: int) => string_add("Some(", string_add(int_to_string((value/2 : int)), ")")),
        Option::None => "None",
    }
  }
}

fn make_optional(flag/3: bool) -> Option[int] {
  if (flag/3 : bool) {
      Option::Some(42)
  } else {
      Option::None
  }
}

fn main() -> unit {
  let point/4: Point = Point { x: 1, y: 2 } in
  let some_number/5: Option[int] = make_optional(true) in
  let none_number/6: Option[int] = make_optional(false) in
  let _ : unit = string_println(show((point/4 : Point))) in
  let _ : unit = string_println(show((some_number/5 : Option[int]))) in
  let _ : unit = string_println(show((none_number/6 : Option[int]))) in
  ()
}