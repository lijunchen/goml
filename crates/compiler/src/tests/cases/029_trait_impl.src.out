Point
Some(42)
None
